File: services/ai-worker-python/fliphouse_worker/stages/_registry.py
"""The stage registry the CLI dispatches over — the 6 Python-backed stages.

Mirrors ``apps/worker-node/src/stages/registry.ts`` PYTHON_STAGES: every stage
the Node worker spawns Python for resolves here, so a wired stage never hits the
fatal ``UNKNOWN_STAGE`` path. ``publish`` is a Node finalizer (it reads the
reframe manifest directly), not a subprocess — so there is no ``store`` stage.
"""

from __future__ import annotations

from collections.abc import Callable

from ._types import StageDeps
from .asr import asr_handler
from .asr_finalize import asr_finalize_handler
from .passthrough import passthrough_handler
from .reframe import reframe_handler
from .score import score_handler
from .transcode import transcode_handler

StageHandler = Callable[[dict], dict]


def build_handlers(deps: StageDeps | None = None) -> dict[str, StageHandler]:
    """Return ``{stage: handler}``. With no deps, build the real env-backed R2 client."""
    if deps is None:
        from .r2 import R2Client

        deps = StageDeps(r2=R2Client.from_env())
    d = deps
    return {
        "transcode": lambda req: transcode_handler(req, d),
        "asr": lambda req: asr_handler(req, d),
        # GigaAM-v3 GPU lane finalizer: validate the parked raw payload → R2 contracts.
        "asr-finalize": lambda req: asr_finalize_handler(req, d),
        "score": lambda req: score_handler(req, d),
        "reframe": lambda req: reframe_handler(req, d),
        "caption": lambda req: passthrough_handler(req, d),
        "banner": lambda req: passthrough_handler(req, d),
    }

File: apps/worker-node/src/stages/registry.ts
import type { Stage } from '@fliphouse/shared';

import { executeStage } from './execute-stage.js';
import type { StageHandler } from './handler-contract.js';

/** Stages driven by a Python subprocess via the generic executeStage body. */
const PYTHON_STAGES: ReadonlySet<Stage> = new Set<Stage>([
  'transcode',
  'asr',
  'score',
  'reframe',
  'caption',
  'banner',
]);

export function isPythonStage(stage: Stage): boolean {
  return PYTHON_STAGES.has(stage);
}

/**
 * Resolve the generic handler for a Python-backed stage. `publish` is a Node
 * finalizer (see publish.ts), not a subprocess stage, so it is routed
 * separately by the worker and rejected here.
 */
export function resolveStageHandler(stage: Stage): StageHandler {
  if (!isPythonStage(stage)) {
    throw new Error(`no generic stage handler for "${stage}"`);
  }
  return executeStage;
}
